storage/replica.go

```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 	batch := r.store.Engine().NewBatch()
 	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
+		// TODO(peter): Could use a "distinct" batch within applySnapshot, but we
```

Comment:
This comment points to the problems with this design: applySnapshot and not its caller should be responsible for deciding whether a distinct batch should be used. It's currently unsafe to use a distinct batch in applySnapshot, but I think #6878 would make it safe (but it may not be useful because applySnapshot will try to read its writes from the batch, so it would have to flush to c++). 

Another possibility: snapshot writes don't need to be atomic with the rest of this method. We use one batch for efficiency, bit it might be better to separate them.